pkg/didcomm/messenger/messenger.go

```diff
@@ -77,16 +81,84 @@ func (m *Messenger) HandleInbound(msg service.DIDCommMsgMap, myDID, theirDID str
 		return fmt.Errorf("threadID: %w", err)
 	}
 
+	if err := m.populateMetadata(thID, msg); err != nil {
+		return fmt.Errorf("with metadata: %w", err)
+	}
+
 	// saves message payload
 	return m.saveRecord(msg.ID(), record{MyDID: myDID, TheirDID: theirDID, ThreadID: thID})
 }
 
+func (m *Messenger) saveMetadata(msg service.DIDCommMsgMap) error {
+	if msg[jsonThread] == nil {
+		return nil
+	}
+
+	thread, ok := msg[jsonThread].(map[string]interface{})
+	if !ok || thread == nil {
+		return nil
+	}
+
+	metadata, ok := thread[jsonMetadata].(map[string]interface{})
+	if !ok || metadata == nil {
+		return nil
+	}
+
+	thID, err := msg.ThreadID()
+	if err != nil {
+		return fmt.Errorf("threadID: %w", err)
+	}
+
+	delete(thread, jsonMetadata)
+
+	// if the thread had only metadata we remove it from the message as well
+	if len(thread) == 0 {
+		delete(msg, jsonThread)
+	}
+
+	return m.saveRecord(fmt.Sprintf(metadataKey, thID), record{Metadata: metadata})
+}
+
+func (m *Messenger) populateMetadata(thID string, msg service.DIDCommMsgMap) error {
+	rec, err := m.getRecord(fmt.Sprintf(metadataKey, thID))
+	if errors.Is(err, storage.ErrDataNotFound) {
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("get record: %w", err)
+	}
+
+	if rec.Metadata == nil {
+		return nil
+	}
+
+	if msg[jsonThread] == nil {
+		msg[jsonThread] = map[string]interface{}{jsonMetadata: rec.Metadata}
```

Comment:
Suggest placing the metadata outside the `~thread`.

Reply:
Done! I have added a comment about how to use it.
```
type MetaData struct {
	Metadata map[string]interface{} `json:"metadata,omitempty"`
}
```